cmake/Modules/FindFLANN.cmake

```diff
@@ -1,63 +1,145 @@
+  set(FLANN_DEFINITIONS ${FLANN_DEFINITIONS} "-DFLANN_STATIC")
 else()
   set(FLANN_RELEASE_NAME flann_cpp)
   set(FLANN_DEBUG_NAME flann_cpp-gd)
 endif()
 
-find_package(PkgConfig QUIET)
-if(FLANN_FIND_VERSION)
-    pkg_check_modules(FLANN flann>=${FLANN_FIND_VERSION})
-else()
-    pkg_check_modules(FLANN flann)
-endif()
-
-if(NOT FLANN_FOUND)
-    find_path(FLANN_INCLUDE_DIR flann/flann.hpp
-              HINTS "${FLANN_ROOT}" "$ENV{FLANN_ROOT}"
-              PATHS "$ENV{PROGRAMFILES}/flann" "$ENV{PROGRAMW6432}/flann"
-              PATH_SUFFIXES include)
-
-    find_library(FLANN_LIBRARY
-                 NAMES ${FLANN_RELEASE_NAME}
-                 HINTS "${FLANN_ROOT}" "$ENV{FLANN_ROOT}"
-                 PATHS "$ENV{PROGRAMFILES}/flann" "$ENV{PROGRAMW6432}/flann"
-                 PATH_SUFFIXES lib)
-
-    find_library(FLANN_LIBRARY_DEBUG
-                 NAMES ${FLANN_DEBUG_NAME} ${FLANN_RELEASE_NAME}
-                 HINTS "${FLANN_ROOT}" "$ENV{FLANN_ROOT}"
-                 PATHS "$ENV{PROGRAMFILES}/flann" "$ENV{PROGRAMW6432}/flann"
-                 PATH_SUFFIXES lib debug/lib)
-
-    if(NOT FLANN_LIBRARY_DEBUG)
-      set(FLANN_LIBRARY_DEBUG ${FLANN_LIBRARY})
-    endif()
+find_path(FLANN_INCLUDE_DIR
+  NAMES
+    flann/flann.hpp
+  HINTS
+    ${PC_FLANN_INCLUDE_DIRS}
+    ${FLANN_ROOT}
+    $ENV{FLANN_ROOT}
+  PATHS
+    $ENV{PROGRAMFILES}/Flann
+    $ENV{PROGRAMW6432}/Flann
+  PATH_SUFFIXES
+    include
+)
 
-    set(FLANN_INCLUDE_DIRS ${FLANN_INCLUDE_DIR})
-    set(FLANN_LIBRARIES optimized ${FLANN_LIBRARY} debug ${FLANN_LIBRARY_DEBUG})
+find_library(FLANN_LIBRARY
+  NAMES
+    ${FLANN_RELEASE_NAME}
+  HINTS
+    ${PC_FLANN_LIBRARY_DIRS}
+    ${FLANN_ROOT}
+    $ENV{FLANN_ROOT}
+  PATHS
+    $ENV{PROGRAMFILES}/Flann
+    $ENV{PROGRAMW6432}/Flann
+  PATH_SUFFIXES
+    lib
+)
 
-    include(FindPackageHandleStandardArgs)
-    find_package_handle_standard_args(FLANN DEFAULT_MSG FLANN_LIBRARY FLANN_INCLUDE_DIR)
+find_library(FLANN_LIBRARY_DEBUG
+  NAMES
+    ${FLANN_DEBUG_NAME}
+  HINTS
+    ${PC_FLANN_LIBRARY_DIRS}
+    ${FLANN_ROOT}
+    $ENV{FLANN_ROOT}
+  PATHS
+    $ENV{PROGRAMFILES}/Flann
+    $ENV{PROGRAMW6432}/Flann
+  PATH_SUFFIXES
+    lib
+)
 
-    mark_as_advanced(FLANN_LIBRARY FLANN_LIBRARY_DEBUG FLANN_INCLUDE_DIR)
+if(FLANN_LIBRARY AND FLANN_LIBRARY_DEBUG)
+  set(FLANN_LIBRARIES optimized ${FLANN_LIBRARY} debug ${FLANN_LIBRARY_DEBUG})
+else()
+  set(FLANN_LIBRARIES ${FLANN_LIBRARY})
 endif()
 
+set(FLANN_INCLUDE_DIRS
+  ${FLANN_INCLUDE_DIR}
+)
+
+include(FindPackageHandleStandardArgs)
+find_package_handle_standard_args(
+  FLANN DEFAULT_MSG
+  FLANN_LIBRARIES FLANN_INCLUDE_DIR
+)
+
 if(FLANN_FOUND)
-  message(STATUS "FLANN found (include: ${FLANN_INCLUDE_DIRS}, lib: ${FLANN_LIBRARIES})")
-  if(FLANN_USE_STATIC)
-    add_definitions(-DFLANN_STATIC)
+  if(NOT TARGET FLANN::FLANN)
+    if (FLANN_USE_STATIC)
+      add_library(FLANN::FLANN STATIC IMPORTED)
+    else()
+      add_library(FLANN::FLANN SHARED IMPORTED)
```

Comment:
Just one comment here: if I'm not wrong, we should add the `INTERFACE` keyword to `add_library` (the same applies to the `STATIC` case). See also the CMake [socumentation](https://cmake.org/cmake/help/v3.5/command/add_library.html?#interface-libraries) about this.

Reply:
i created a ticket at CMake repo for [this](https://gitlab.kitware.com/cmake/cmake/issues/18975), because i don't see when to use Interface Imported Target and when to use Imported Target.

Reply:
I got an answer: Imported target without `INTERFACE` keyword is correct.

Reply:
Ok, thanks!